File: scala/bob/src/main/scala/bob.scala
class Bob {

  def hey(greeting: String): String = {
    if (isYelled(greeting)) "Whoa, chill out!"
    else if (isQuestion(greeting)) "Sure."
    else if (isWhiteSpace(greeting)) "Fine. Be that way!"
    else "Whatever."
  }

  def isYelled(greeting: String): Boolean = {
    val hasLetter: Boolean = greeting.exists(_.isLetter) 
    val isAllCaps: Boolean = greeting.toUpperCase == greeting
    hasLetter && isAllCaps
  }

  def isQuestion(greeting: String): Boolean = greeting.endsWith("?")

  def isWhiteSpace(greeting: String): Boolean = greeting.trim.isEmpty

}

File: python/bob/bob.py
# -*- coding: utf-8 -*-
#!/usr/bin/env python
import re


def is_yelled(greeting):
    has_alpha_char = bool(re.search(ur'[A-Z]', greeting))
    is_all_caps = greeting.isupper()
    return has_alpha_char and is_all_caps

def is_asked(greeting):
    # Ensure greeting is not empty and check last char for question mark
    return False if not greeting else greeting[-1] == '?'

def is_whitespace(greeting):
    # local greeting is just hey's greeting with leading and trailing whitespace
    # removed --> will be empty string if hey's greeting was all whitespace
    return True if not greeting else False

def hey(greeting):
    ugreeting = greeting.strip()
    if is_yelled(ugreeting):
        return u"Whoa, chill out!"
    elif is_asked(ugreeting):
        return u"Sure."
    elif is_whitespace(ugreeting):
        return u"Fine. Be that way!"
    else:
        return u"Whatever."
